node/core/approval-voting/src/lib.rs

```diff
@@ -768,25 +796,64 @@ async fn handle_actions(
 				let block_hash = indirect_cert.block_hash;
 				let validator_index = indirect_cert.validator;
 
-				ctx.send_unbounded_message(ApprovalDistributionMessage::DistributeAssignment(
-					indirect_cert,
-					candidate_index,
-				).into());
+				match state.approvals_cache.get(&candidate_hash) {
```

Comment:
👍 

I'd rather have `approvals_cache` go in the parameters to the function than the state. It is true that this function is getting unwieldy so refactoring them to a `struct HandleActionsParams<'a>` would be welcome.